### ISSUE TITLE
Performance 5x degradation after converting numpy to tf.function equivalent

### ISSUE THREAD

**Issue by emadboctorx** [tensorflow/tensorflow#46126]
<em>Please make sure that this is an issue related to performance of TensorFlow.
As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:performance_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Custom code
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu (Google colab) & macOs 11.1
- TensorFlow installed from (source or binary): pip installation.
- TensorFlow version (use command below): 2.3.1 / 2.4
- Python version: 3.6+
- CUDA/cuDNN version: 10.1
- GPU model and memory: Varies

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

I converted this function, that performs gradient updates for a DQN agent. It takes a batch of numpy arrays, calculates target values and updates gradients using `model.fit()` for 1 epoch, it usually processes 90-100 frames per second on google colab.

    def update(self, batch):
        """
        Update gradients given a batch.
        Args:
            batch: A batch of observations in the form of
                [[states], [actions], [rewards], [dones], [next states]]
    
        Returns:
            None
        """
        states, actions, rewards, dones, new_states = batch
        q_states = self.main_model.predict(states)
        if self.double:
            new_state_actions = np.argmax(self.main_model.predict(new_states), 1)
            new_state_q_values = self.target_model.predict(new_states)
            new_state_values = new_state_q_values[
                np.arange(self.batch_size), new_state_actions
            ]
        else:
            new_state_values = self.target_model.predict(new_states).max(1)
        new_state_values[dones] = 0
        target_values = np.copy(q_states)
        target_value_update = new_state_values * self.gamma ** self.n_steps + rewards
        state_action_values = target_values[np.arange(self.batch_size), actions]
        target_values[np.arange(self.batch_size), actions] = target_value_update
        self.main_model.fit(states, target_values, verbose=0)
        if self.buffer.priorities:
            squared_loss = (state_action_values - target_value_update) ** 2
            priorities = (
                self.buffer.current_weights * squared_loss + self.buffer.priority_bias
            )
            self.buffer.update_priorities(priorities)

**After the conversion:**

    @tf.function
    def get_targets(self, batch):
        """
        Get target values for gradient updates.
        Args:
            batch: A batch of observations in the form of
                [[states], [actions], [rewards], [dones], [next states]]
        Returns:
            None
        """
        states, actions, rewards, dones, new_states = batch
        q_states = self.main_model(states)
        if self.double:
            new_state_actions = tf.argmax(self.main_model(new_states), 1)
            new_state_q_values = self.target_model(new_states)
            a = self.get_action_indices(new_state_actions)
            new_state_values = tf.gather_nd(new_state_q_values, a)
        else:
            new_state_values = tf.reduce_max(self.target_model(new_states), axis=1)
        new_state_values *= tf.cast(~dones, tf.float32)
        target_values = tf.identity(q_states)
        target_value_update = new_state_values * self.gamma ** self.n_steps + tf.cast(
            rewards, tf.float32
        )
        indices = self.get_action_indices(actions)
        state_action_values = tf.gather_nd(target_values, indices)
        tf.tensor_scatter_nd_update(target_values, indices, target_value_update)
        if self.buffer.priorities:
            squared_loss = (state_action_values - target_value_update) ** 2
            priorities = (
                self.buffer.current_weights * squared_loss + self.buffer.priority_bias
            )
            self.buffer.update_priorities(priorities)
        return target_values

And I get `model.fit()` outside the method which results in a severe performance degradation ~= 20 frames per seconds.

**Describe the expected behavior**

**Standalone code to reproduce the issue**
Here's the colab [notebook](https://colab.research.google.com/drive/1PNSURh-Hhd2CtQ5vn82lG0rsJjYkYPqR?usp=sharing) for the original version(numpy, no `tf.function`)

Here's the less performant / worse version colab [notebook](https://colab.research.google.com/drive/1qSzJSepVSYyA3dRpdpgHo2cRWeOg2nSJ?usp=sharing)

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by Saduf2019** [comment]
@emadboctorx 
I ran the code shared in the notebook but do not face any errors please find [gist here](https://colab.research.google.com/gist/Saduf2019/9ded64a8ca6d8fab5827772d5afa7729/untitled491.ipynb).

**Comment by emadboctorx** [comment]
@Saduf2019 That's because you ran the good version, now try running the bad [version](https://colab.research.google.com/drive/1qSzJSepVSYyA3dRpdpgHo2cRWeOg2nSJ?usp=sharing) which I included in the description, you should experience a severe slowdown (no errors)

**Comment by WindQAQ** [comment]
@emadboctorx Try this. https://colab.research.google.com/drive/1JjEcnSnGCd_Jkqw-6LYNtJ7PcEs0FTgH?usp=sharing

**Comment by emadboctorx** [comment]
@WindQAQ I remember telling you a couple of days back you're a life saver! what was the problem? the speed almost doubled with your version, can you tell me how did you fix it, in order for me to follow the same steps in the future?

**Comment by WindQAQ** [comment]
@emadboctorx  I find two points

1. `tf.keras.Model.train_on_batch()` introduces incredible overhead in your case. Using custom training loop can be much faster.
2. In `get_action`, you are using `tf.keras.Model.predict` on a single batch of data. Using `tf.keras.Model.predict_on_batch` is much faster.

I haven't digged into why though.

**Comment by emadboctorx** [comment]
Thanks anyway, that totally fixes the problem. Also you added this line in the model creation method, I never have seen it before.

```
    model.call = tf.function(model.call)
```

**Comment by WindQAQ** [comment]
It's used to decorate the call function with `tf.function`. It does not matter if your outer function is decorated with `tf.function` though. If you are using `{train,test,predict}` or `{train,test,predict}_on_batch`, it is natively decorated with `tf.function` because your are compiling a model with `run_eagerly=False`. However, nested `tf.function` might or might not improve the performance. It depends on if the function is inlined.

**Comment by WindQAQ** [comment]
The problem is that `train_on_batch` will generate a `tf.data.Dataset` placed on CPU. Therefore, it will first create a tensor on GPU, and then transfer to GPU for training. The data transferring cost from host to device is not negligible especially for large tensor (in your case (32, 84, 84, 1)).

https://colab.research.google.com/drive/1s6O4MyfqyCFI8uEM6JlTVNF18BLvYfka?usp=sharing

**Comment by emadboctorx** [comment]
That's almost 3x the speed of `model.train_on_batch()`. And yes I'm familiar with `tf.data.Dataset` I totally rely on the api for training supervised learning models using tfrecords, it's ultra-convenient, flexible and efficient, but unfortunately, this cannot be applied to online learning which is the case here, because the dataset constantly changes as the training progresses. I want to improve the performance and maybe parallelize the training process, I'm trying to figure out how.
